File: backend/app/schemas/green.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class GreenItemInput(BaseModel):
    """单条活动数据：code 为因子库编码（可空，缺失时按 factor_missing 处理）。"""

    code: str = ""
    name: str = ""
    quantity: float = Field(gt=0)
    unit: str = ""
    note: str = ""


class GreenAnalyzeForm(BaseModel):
    project_id: str
    area_m2: float | None = Field(default=None, gt=0)
    scope: str = ""
    materials: list[GreenItemInput] = Field(default_factory=list)
    transport: list[GreenItemInput] = Field(default_factory=list)
    energy: list[GreenItemInput] = Field(default_factory=list)


class CarbonItemRead(BaseModel):
    category: str  # material | energy | transport
    stage: str  # A1-A3 | A4 | A5
    stage_name: str
    code: str
    name: str
    unit: str
    quantity: float
    emission_factor: float | None
    factor_unit: str
    emission: float  # tCO2e
    factor_source: str
    verified: bool = False
    factor_missing: bool = False
    note: str = ""


class CarbonStageSummary(BaseModel):
    stage: str
    stage_name: str
    emission: float
    share: float  # 0~1，占总量比例
    items_count: int


class CarbonContributor(BaseModel):
    code: str
    name: str
    category: str
    stage: str
    emission: float
    share: float


class CarbonAnalysisResponse(BaseModel):
    analysis_id: str
    project_id: str
    project_name: str
    created_at: str
    area_m2: float | None
    scope: str
    total_emission: float
    unit: str = "tCO2e"
    intensity: float | None  # tCO2e/m2
    stages: list[CarbonStageSummary]
    items: list[CarbonItemRead]
    top_contributors: list[CarbonContributor]
    suggestions: list[str]
    factor_version: str
    has_unverified_factors: bool
    factor_warnings: list[str]
    report_preview: str
    is_simulated: bool


class CarbonAnalysisSummary(BaseModel):
    analysis_id: str
    project_id: str
    project_name: str
    area_m2: float | None
    scope: str
    total_emission: float
    is_simulated: bool
    has_unverified_factors: bool
    created_at: str


class BenchmarkItem(BaseModel):
    """跨项目碳强度对标中的单个项目。"""

    rank: int
    project_id: str
    project_name: str
    intensity: float | None  # tCO2e/m²
    z: float  # (intensity-mean)/std，负值=低于均值=更优
    better_than_pct: float  # 0~100，池内严格劣于当前的项目占比


class GreenBenchmark(BaseModel):
    """按用户可见项目池，用最新一次核算的面积强度做 z-score 排名。"""

    available: bool
    reason: str | None  # 不可用原因，如「样本不足 2 个项目」/「标准差为 0」
    count: int
    metric: str = "intensity"
    unit: str = "tCO2e/m²"
    mean: float | None
    std: float | None
    current: BenchmarkItem | None  # 当前项目在榜内的位置
    items: list[BenchmarkItem]  # 按 intensity 升序，rank 1 最优


class FactorRead(BaseModel):
    code: str
    category: str
    name: str
    unit: str
    factor: float
    factor_unit: str
    source: str
    year: int | None = None
    verified: bool = False
    note: str = ""


class ReferenceMetricRead(BaseModel):
    """单条真实公开数据（如中国建筑 2024 年度 ESG/年报披露）。"""

    code: str
    name: str
    value: str
    unit: str
    year: int | None = None
    source: str
    note: str = ""


class ReferenceGroupRead(BaseModel):
    category: str
    name: str
    items: list[ReferenceMetricRead]


class GreenReference(BaseModel):
    """真实公开数据参考库（独立于本项目 z-score 对标，供对标参考）。"""

    version: str
    updated_at: str
    source_note: str
    groups: list[ReferenceGroupRead]


# ---------- 四节一环保评估 ----------


class GreenMetricInput(BaseModel):
    """单个评估指标输入。value 可空：缺失的指标该维度按 0 计并标记模拟。"""

    key: str
    value: float | None = Field(default=None, ge=0)


class DimensionInput(BaseModel):
    """一个维度的子指标输入。dimension 取值 material | water | energy | land | env。"""

    dimension: str
    metrics: list[GreenMetricInput] = Field(default_factory=list)


class GreenAssessmentForm(BaseModel):
    project_id: str
    title: str = ""
    area_m2: float | None = Field(default=None, ge=0)
    dimensions: list[DimensionInput] = Field(default_factory=list)


class MetricScore(BaseModel):
    key: str
    name: str
    value: float | None
    target: float
    direction: str  # higher | lower
    score: float  # 0~100


class DimensionScore(BaseModel):
    dimension: str
    name: str
    score: float  # 子指标得分均值 0~100
    metrics: list[MetricScore]


class GreenAssessmentResponse(BaseModel):
    assessment_id: str
    project_id: str
    project_name: str
    title: str
    area_m2: float | None
    total_score: float
    level: str  # 不合格 | 合格 | 优良 | 优秀
    dimensions: list[DimensionScore]
    is_simulated: bool
    report_preview: str
    created_at: str


class GreenAssessmentSummary(BaseModel):
    assessment_id: str
    project_id: str
    project_name: str
    title: str
    total_score: float
    level: str
    is_simulated: bool
    created_at: str


# ---------- 环保监测台账 ----------


class GreenEnvRecordForm(BaseModel):
    """当日环保监测读数，均可空。同一项目+日期重复提交为 upsert（幂等重录）。"""

    project_id: str
    record_date: date
    pm25: float | None = Field(default=None, ge=0)
    pm10: float | None = Field(default=None, ge=0)
    tsp: float | None = Field(default=None, ge=0)
    noise_day_db: float | None = Field(default=None, ge=0)
    noise_night_db: float | None = Field(default=None, ge=0)
    cod_mg: float | None = Field(default=None, ge=0)
    ss_mg: float | None = Field(default=None, ge=0)
    ph: float | None = Field(default=None, ge=0)
    solid_waste_t: float | None = Field(default=None, ge=0)


class EnvThresholdRead(BaseModel):
    """UI 展示用阈值常量。rule=above 超上限告警；rule=range 越界告警（如 pH）。"""

    key: str
    name: str
    unit: str
    rule: str  # above | range
    limit: float | None = None
    min: float | None = None
    max: float | None = None


class EnvAlertRead(BaseModel):
    key: str
    name: str
    value: float
    rule: str
    limit: float | None = None
    min: float | None = None
    max: float | None = None


class EnvRecordRead(BaseModel):
    record_id: str
    project_id: str
    project_name: str
    record_date: str
    pm25: float | None
    pm10: float | None
    tsp: float | None
    noise_day_db: float | None
    noise_night_db: float | None
    cod_mg: float | None
    ss_mg: float | None
    ph: float | None
    solid_waste_t: float | None
    has_alerts: bool
    alerts: list[EnvAlertRead]
    created_at: str


# ---------- 碳排趋势与目标 ----------


class GreenTargetForm(BaseModel):
    project_id: str
    target_intensity: float | None = Field(default=None, gt=0)
    note: str = ""


class GreenTargetRead(BaseModel):
    project_id: str
    target_intensity: float | None
    note: str
    updated_at: str


class GreenTrendPoint(BaseModel):
    created_at: str
    total_emission: float
    area_m2: float
    intensity: float


class GreenTrendCurrent(BaseModel):
    intensity: float | None
    target_intensity: float | None
    grade: str  # 达标 | 临界 | 超标 | 未设目标
    gap_pct: float | None  # 相对目标百分比，负值=优于目标


class GreenTrendResponse(BaseModel):
    project_id: str
    project_name: str
    points: list[GreenTrendPoint]
    current: GreenTrendCurrent


# ---------- AI 优化建议 ----------


class GreenAdviceForm(BaseModel):
    project_id: str
    source_type: str  # carbon | assessment
    analysis_id: str | None = None
    assessment_id: str | None = None


class GreenAdviceRead(BaseModel):
    advice: str
    is_simulated: bool
    source_type: str
    generated_at: str

File: frontend/src/types/green.ts
export interface GreenItemInput {
  code: string
  name: string
  quantity: number
  unit: string
  note?: string
}

export interface GreenAnalyzeForm {
  project_id: string
  area_m2: number | null
  scope: string
  materials: GreenItemInput[]
  transport: GreenItemInput[]
  energy: GreenItemInput[]
}

export interface CarbonItem {
  category: 'material' | 'energy' | 'transport'
  stage: string
  stage_name: string
  code: string
  name: string
  unit: string
  quantity: number
  emission_factor: number | null
  factor_unit: string
  emission: number
  factor_source: string
  verified: boolean
  factor_missing: boolean
  note: string
}

export interface CarbonStage {
  stage: string
  stage_name: string
  emission: number
  share: number
  items_count: number
}

export interface CarbonContributor {
  code: string
  name: string
  category: string
  stage: string
  emission: number
  share: number
}

export interface CarbonAnalysisResult {
  analysis_id: string
  project_id: string
  project_name: string
  created_at: string
  area_m2: number | null
  scope: string
  total_emission: number
  unit: string
  intensity: number | null
  stages: CarbonStage[]
  items: CarbonItem[]
  top_contributors: CarbonContributor[]
  suggestions: string[]
  factor_version: string
  has_unverified_factors: boolean
  factor_warnings: string[]
  report_preview: string
  is_simulated: boolean
}

export interface CarbonAnalysisSummary {
  analysis_id: string
  project_id: string
  project_name: string
  area_m2: number | null
  scope: string
  total_emission: number
  is_simulated: boolean
  has_unverified_factors: boolean
  created_at: string
}

export interface GreenFactor {
  code: string
  category: 'material' | 'energy' | 'transport'
  name: string
  unit: string
  factor: number
  factor_unit: string
  source: string
  year: number | null
  verified: boolean
  note: string
}

export interface BenchmarkItem {
  rank: number
  project_id: string
  project_name: string
  intensity: number | null
  z: number
  better_than_pct: number
}

export interface GreenBenchmark {
  available: boolean
  reason: string | null
  count: number
  metric: string
  unit: string
  mean: number | null
  std: number | null
  current: BenchmarkItem | null
  items: BenchmarkItem[]
}

export interface ReferenceMetric {
  code: string
  name: string
  value: string
  unit: string
  year: number | null
  source: string
  note: string
}

export interface ReferenceGroup {
  category: string
  name: string
  items: ReferenceMetric[]
}

export interface GreenReference {
  version: string
  updated_at: string
  source_note: string
  groups: ReferenceGroup[]
}

// ---------- 四节一环保评估 ----------

export interface GreenMetricInput {
  key: string
  value: number | null
}

export interface DimensionInput {
  dimension: 'material' | 'water' | 'energy' | 'land' | 'env'
  metrics: GreenMetricInput[]
}

export interface GreenAssessmentForm {
  project_id: string
  title?: string
  area_m2?: number | null
  dimensions: DimensionInput[]
}

export interface MetricScore {
  key: string
  name: string
  value: number | null
  target: number
  direction: 'higher' | 'lower'
  score: number
}

export interface DimensionScore {
  dimension: string
  name: string
  score: number
  metrics: MetricScore[]
}

export interface GreenAssessmentResult {
  assessment_id: string
  project_id: string
  project_name: string
  title: string
  area_m2: number | null
  total_score: number
  level: string
  dimensions: DimensionScore[]
  is_simulated: boolean
  report_preview: string
  created_at: string
}

export interface GreenAssessmentSummary {
  assessment_id: string
  project_id: string
  project_name: string
  title: string
  total_score: number
  level: string
  is_simulated: boolean
  created_at: string
}

// ---------- 环保监测台账 ----------

export interface GreenEnvRecordForm {
  project_id: string
  record_date: string
  pm25: number | null
  pm10: number | null
  tsp: number | null
  noise_day_db: number | null
  noise_night_db: number | null
  cod_mg: number | null
  ss_mg: number | null
  ph: number | null
  solid_waste_t: number | null
}

export interface EnvThreshold {
  key: string
  name: string
  unit: string
  rule: 'above' | 'range'
  limit: number | null
  min: number | null
  max: number | null
}

export interface EnvAlert {
  key: string
  name: string
  value: number
  rule: string
  limit: number | null
  min: number | null
  max: number | null
}

export interface EnvRecord {
  record_id: string
  project_id: string
  project_name: string
  record_date: string
  pm25: number | null
  pm10: number | null
  tsp: number | null
  noise_day_db: number | null
  noise_night_db: number | null
  cod_mg: number | null
  ss_mg: number | null
  ph: number | null
  solid_waste_t: number | null
  has_alerts: boolean
  alerts: EnvAlert[]
  created_at: string
}

// ---------- 碳排趋势与目标 ----------

export interface GreenTargetForm {
  project_id: string
  target_intensity: number | null
  note?: string
}

export interface GreenTarget {
  project_id: string
  target_intensity: number | null
  note: string
  updated_at: string
}

export interface GreenTrendPoint {
  created_at: string
  total_emission: number
  area_m2: number
  intensity: number
}

export interface GreenTrendCurrent {
  intensity: number | null
  target_intensity: number | null
  grade: string
  gap_pct: number | null
}

export interface GreenTrend {
  project_id: string
  project_name: string
  points: GreenTrendPoint[]
  current: GreenTrendCurrent
}

// ---------- AI 优化建议 ----------

export interface GreenAdviceForm {
  project_id: string
  source_type: 'carbon' | 'assessment'
  analysis_id?: string | null
  assessment_id?: string | null
}

export interface GreenAdvice {
  advice: string
  is_simulated: boolean
  source_type: string
  generated_at: string
}
